# **Introduction**

We will start by mentioning our intuition about the data after a fast look. We can see that the school performs interviews and tests for the applicants. We believe the scores of these tests and interviews are of a great contribution to the decision process. This is our belief that we are yet to confirm or reject based on our analysis further down the road.

We will begin by understanding and exploring the data. Then we will perform data cleaning and data wrangling to organize our data and keep the useful data only. Then we will perform correlation analysis to see whether certain features have strong correlations between each other or not. We will also see if some features have high correlations with the decision of enrollment or not and this can help us decide on the factors contributing to the decision process.

After the data analysis, we will create models that can help predict the decision processes. The types and functions of the models will depend on how our data analysis goes. At the moment, we are thinking of regression models to predict the interview scores, only if the scores have high correlations with the decision, and models to predict the decision, but the chosen model will depend on how our data analysis goes.

# **Descriptive Analytics**

## **Basic data exploration**

We will start by data exploration. First let's import the libraries needed throught the notebook and then read the data.

In [191]:
import numpy as np  
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib as mpl
import plotly.express as px
import plotly.graph_objs as pgo 
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import LeakyReLU
from tensorflow.keras.optimizers import Adam 
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

In [192]:
file_name = "C:/Users/moham/OneDrive/Desktop/train_International-School-Dataset.csv"
#file_name = "C:/Users/HP/Downloads/train_International-School-Dataset.csv"
data = pd.read_csv(file_name)

Let's take a quick look at the data.

In [193]:
data

app_id    job_name  job_exp_yrs address house_type  annual_salary  \
0      126229306   Assistant          1.0     YUU     rental        46000.0   
1       76692947       Nurse          1.0     YUU     rental       152000.0   
2       60624620    engineer          5.0     NXT     rental       115000.0   
3      129161039     manager          2.0     YDC     rental        25000.0   
4       90518461       Other          6.0     LJM   mortgage        80000.0   
...          ...         ...          ...     ...        ...            ...   
33995   57383315     Analyst          5.0     YUU     rental       105000.0   
33996   66364198   Marketing          3.0     RQX   mortgage       118299.0   
33997   23412190     teacher          6.0     UWH   mortgage        56000.0   
33998  106575453     manager          2.0     CAA   mortgage       107100.0   
33999  115722215  Consultant          1.0     XSO   mortgage       125000.0   

      salary_verification  n_semesters  prev_grade_percent  year  ...  \
0                       V            1               92.65  2017  ...   
1                       V            1               93.51  2016  ...   
2                       V            2               88.47  2015  ...   
3                      NV            1               89.58  2018  ...   
4                       V            1               78.51  2016  ...   
...                   ...          ...                 ...   ...  ...   
33995                   V            1               91.82  2015  ...   
33996                   V            2               85.52  2015  ...   
33997                 VbP            2               85.01  2014  ...   
33998                 VbP            1               85.01  2017  ...   
33999                 VbP            1               92.93  2017  ...   

       num_children  credit_limit  credit_limit_ex_mort spouse_income  \
0                 3       68039.0               35198.0      110750.0   
1                 5       24000.0               11704.0           NaN   
2                 5       67122.0               41004.0           NaN   
3                 2       12000.0                4065.0           NaN   
4                 2      211158.0               32595.0           NaN   
...             ...           ...                   ...           ...   
33995             2       60500.0               28742.0           NaN   
33996             2      470072.0               69265.0           NaN   
33997             4      279870.0               38038.0           NaN   
33998             4      287972.0               17905.0           NaN   
33999             5      432200.0               40925.0           NaN   

       spouse_fees/income spouse_income_verification      decision  \
0                   27.97                          V  not enrolled   
1                     NaN                        NaN      enrolled   
2                     NaN                        NaN      enrolled   
3                     NaN                        NaN  not enrolled   
4                     NaN                        NaN      enrolled   
...                   ...                        ...           ...   
33995                 NaN                        NaN      enrolled   
33996                 NaN                        NaN      enrolled   
33997                 NaN                        NaN      enrolled   
33998                 NaN                        NaN      enrolled   
33999                 NaN                        NaN      enrolled   

       mental_iv_score  parent_iv_score  social_iv_score  
0             0.087638         8.423544        32.629941  
1            25.328459        12.940269        21.888056  
2             7.141356         7.539012        20.730058  
3            34.474615        22.732014         1.076913  
4            29.070282         5.226745        13.420388  
...                ...              ...              ...  
33995        16.907394        64.581258         1.609

In [194]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34000 entries, 0 to 33999
Data columns (total 26 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   app_id                      34000 non-null  int64  
 1   job_name                    30102 non-null  object 
 2   job_exp_yrs                 34000 non-null  float64
 3   address                     34000 non-null  object 
 4   house_type                  33993 non-null  object 
 5   annual_salary               34000 non-null  float64
 6   salary_verification         34000 non-null  object 
 7   n_semesters                 34000 non-null  int64  
 8   prev_grade_percent          34000 non-null  float64
 9   year                        34000 non-null  int64  
 10  fees/income                 33997 non-null  float64
 11  down_pymnt                  34000 non-null  float64
 12  fees                        34000 non-null  float64
 13  spouse_has_job              340

In [195]:
data.describe()

app_id   job_exp_yrs  annual_salary   n_semesters  \
count  3.400000e+04  34000.000000   3.400000e+04  34000.000000   
mean   6.131298e+07      4.079294   9.820583e+04      1.430647   
std    3.771724e+07      2.606378   7.353028e+04      0.495174   
min    4.058310e+05      1.000000   0.000000e+00      1.000000   
25%    3.089923e+07      2.000000   6.500000e+04      1.000000   
50%    6.173726e+07      4.000000   8.500000e+04      1.000000   
75%    8.847571e+07      6.000000   1.150000e+05      2.000000   
max    1.459724e+08      9.000000   7.446395e+06      2.000000   

       prev_grade_percent          year   fees/income    down_pymnt  \
count        34000.000000  34000.000000  33997.000000  34000.000000   
mean            85.424298   2015.244735     15.538769  12163.304118   
std              5.407398      1.401055     10.217856  10707.437714   
min             69.010000   2012.000000      0.000000      0.000000   
25%             82.140000   2014.000000      9.510000      0.000000   
50%             86.010000   2015.000000     14.540000  16000.000000   
75%             89.250000   2016.000000     20.410000  20000.000000   
max             94.690000   2018.000000    740.960000  32000.000000   

               fees  total_balance  num_children  credit_limit  \
count  34000.000000   3.400000e+04  34000.000000  3.400000e+04   
mean   25950.638235   1.577643e+05      3.128441  1.970130e+05   
std     5468.639711   1.834909e+05      1.481278  2.013486e+05   
min    20000.000000   0.000000e+00      0.000000  0.000000e+00   
25%    21000.000000   3.195675e+04      2.000000  5.696950e+04   
50%    24950.000000   7.143400e+04      3.000000  1.158680e+05   
75%    30000.000000   2.389780e+05      4.000000  2.863752e+05   
max    40000.000000   4.772549e+06      5.000000  4.932524e+06   

       credit_limit_ex_mort  spouse_income  spouse_fees/income  \
count          34000.000000    1294.000000         1294.000000   
mean           43389.425441  125014.670889           17.476236   
std            35935.186485   58939.673848            7.327846   
min                0.000000   40000.000000            0.230000   
25%            23465.250000   87500.000000           12.037500   
50%            36382.000000  110250.000000           17.110000   
75%            53274.000000  148000.000000           22.435000   
max           897835.000000  642000.000000           39.900000   

       mental_iv_score  parent_iv_score  social_iv_score  
count     34000.000000     34000.000000     34000.000000  
mean         20.843090        20.887305        21.022355  
std          16.648062        16.696663        16.792907  
min           0.000667         0.000027         0.000666  
25%           7.406833         7.395480         7.485895  
50%          16.886263        16.937863        16.980574  
75%          30.683819        30.723531        30.929671  
max          87.851114        88.246603        88.873714

## **Data Cleaning**

We will now start data cleaning. Let's check for duplicates.

### Checking for duplicates

In [196]:
data['app_id'].duplicated().sum()

0

In [197]:
temp_data = data.drop(columns = "app_id")
temp_data

job_name  job_exp_yrs address house_type  annual_salary  \
0       Assistant          1.0     YUU     rental        46000.0   
1           Nurse          1.0     YUU     rental       152000.0   
2        engineer          5.0     NXT     rental       115000.0   
3         manager          2.0     YDC     rental        25000.0   
4           Other          6.0     LJM   mortgage        80000.0   
...           ...          ...     ...        ...            ...   
33995     Analyst          5.0     YUU     rental       105000.0   
33996   Marketing          3.0     RQX   mortgage       118299.0   
33997     teacher          6.0     UWH   mortgage        56000.0   
33998     manager          2.0     CAA   mortgage       107100.0   
33999  Consultant          1.0     XSO   mortgage       125000.0   

      salary_verification  n_semesters  prev_grade_percent  year  fees/income  \
0                       V            1               92.65  2017        17.48   
1                       V            1               93.51  2016         2.39   
2                       V            2               88.47  2015        12.95   
3                      NV            1               89.58  2018         9.75   
4                       V            1               78.51  2016        15.17   
...                   ...          ...                 ...   ...          ...   
33995                   V            1               91.82  2015        11.71   
33996                   V            2               85.52  2015        13.28   
33997                 VbP            2               85.01  2014        18.88   
33998                 VbP            1               85.01  2017         3.61   
33999                 VbP            1               92.93  2017         6.56   

       ...  num_children  credit_limit credit_limit_ex_mort  spouse_income  \
0      ...             3       68039.0              35198.0       110750.0   
1      ...             5       24000.0              11704.0            NaN   
2      ...             5       67122.0              41004.0            NaN   
3      ...             2       12000.0               4065.0            NaN   
4      ...             2      211158.0              32595.0            NaN   
...    ...           ...           ...                  ...            ...   
33995  ...             2       60500.0              28742.0            NaN   
33996  ...             2      470072.0              69265.0            NaN   
33997  ...             4      279870.0              38038.0            NaN   
33998  ...             4      287972.0              17905.0            NaN   
33999  ...             5      432200.0              40925.0            NaN   

      spouse_fees/income  spouse_income_verification      decision  \
0                  27.97                           V  not enrolled   
1                    NaN                         NaN      enrolled   
2                    NaN                         NaN      enrolled   
3                    NaN                         NaN  not enrolled   
4                    NaN                         NaN      enrolled   
...                  ...                         ...           ...   
33995                NaN                         NaN      enrolled   
33996                NaN                         NaN      enrolled   
33997                NaN                         NaN      enrolled   
33998                NaN                         NaN      enrolled   
33999                NaN                         NaN      enrolled   

       mental_iv_score  parent_iv_score  social_iv_score  
0             0.087638         8.423544        32.629941  
1            25.328459        12.940269        21.888056  
2             7.141356         7.539012        20.730058  
3            34.474615        22.732014         1.076913  
4            29.070282         5.226745        13.420388  
...                ...              ...              ...  
33995        16.907394      

In [198]:
temp_data.duplicated().sum()

0

In [199]:
del(temp_data)

No duplicates were found at all. Next, we will check for missing values.

### Checking for missing values

In [200]:
data.isnull().any()

app_id                        False
job_name                       True
job_exp_yrs                   False
address                       False
house_type                     True
annual_salary                 False
salary_verification           False
n_semesters                   False
prev_grade_percent            False
year                          False
fees/income                    True
down_pymnt                    False
fees                          False
spouse_has_job                False
total_balance                 False
payment_method                False
num_children                  False
credit_limit                  False
credit_limit_ex_mort          False
spouse_income                  True
spouse_fees/income             True
spouse_income_verification     True
decision                      False
mental_iv_score               False
parent_iv_score               False
social_iv_score               False
dtype: bool

Job name, house type, fees / income, spouse income, spouse fees / income, spouse income verification all have missing values. We will start by analysing the missing values of house type. 

In [201]:
data['house_type'].isnull().sum()

7

There are only 7 data points in the data that have N/A house type. we can remove them because they will not be significant in the classification and if we included them we will have to encode using one hot encoding and add a whole column just for these 7 data points. So, we will remove them to make things easier.

In [202]:
data = data[data['house_type'].isnull() == False]
data

app_id    job_name  job_exp_yrs address house_type  annual_salary  \
0      126229306   Assistant          1.0     YUU     rental        46000.0   
1       76692947       Nurse          1.0     YUU     rental       152000.0   
2       60624620    engineer          5.0     NXT     rental       115000.0   
3      129161039     manager          2.0     YDC     rental        25000.0   
4       90518461       Other          6.0     LJM   mortgage        80000.0   
...          ...         ...          ...     ...        ...            ...   
33995   57383315     Analyst          5.0     YUU     rental       105000.0   
33996   66364198   Marketing          3.0     RQX   mortgage       118299.0   
33997   23412190     teacher          6.0     UWH   mortgage        56000.0   
33998  106575453     manager          2.0     CAA   mortgage       107100.0   
33999  115722215  Consultant          1.0     XSO   mortgage       125000.0   

      salary_verification  n_semesters  prev_grade_percent  year  ...  \
0                       V            1               92.65  2017  ...   
1                       V            1               93.51  2016  ...   
2                       V            2               88.47  2015  ...   
3                      NV            1               89.58  2018  ...   
4                       V            1               78.51  2016  ...   
...                   ...          ...                 ...   ...  ...   
33995                   V            1               91.82  2015  ...   
33996                   V            2               85.52  2015  ...   
33997                 VbP            2               85.01  2014  ...   
33998                 VbP            1               85.01  2017  ...   
33999                 VbP            1               92.93  2017  ...   

       num_children  credit_limit  credit_limit_ex_mort spouse_income  \
0                 3       68039.0               35198.0      110750.0   
1                 5       24000.0               11704.0           NaN   
2                 5       67122.0               41004.0           NaN   
3                 2       12000.0                4065.0           NaN   
4                 2      211158.0               32595.0           NaN   
...             ...           ...                   ...           ...   
33995             2       60500.0               28742.0           NaN   
33996             2      470072.0               69265.0           NaN   
33997             4      279870.0               38038.0           NaN   
33998             4      287972.0               17905.0           NaN   
33999             5      432200.0               40925.0           NaN   

       spouse_fees/income spouse_income_verification      decision  \
0                   27.97                          V  not enrolled   
1                     NaN                        NaN      enrolled   
2                     NaN                        NaN      enrolled   
3                     NaN                        NaN  not enrolled   
4                     NaN                        NaN      enrolled   
...                   ...                        ...           ...   
33995                 NaN                        NaN      enrolled   
33996                 NaN                        NaN      enrolled   
33997                 NaN                        NaN      enrolled   
33998                 NaN                        NaN      enrolled   
33999                 NaN                        NaN      enrolled   

       mental_iv_score  parent_iv_score  social_iv_score  
0             0.087638         8.423544        32.629941  
1            25.328459        12.940269        21.888056  
2             7.141356         7.539012        20.730058  
3            34.474615        22.732014         1.076913  
4            29.070282         5.226745        13.420388  
...                ...              ...              ...  
33995        16.907394        64.581258         1.609

In [203]:
data.loc[(data['spouse_has_job'] == 'Yes') & (data['spouse_income'].isnull())]

Empty DataFrame
Columns: [app_id, job_name, job_exp_yrs, address, house_type, annual_salary, salary_verification, n_semesters, prev_grade_percent, year, fees/income, down_pymnt, fees, spouse_has_job, total_balance, payment_method, num_children, credit_limit, credit_limit_ex_mort, spouse_income, spouse_fees/income, spouse_income_verification, decision, mental_iv_score, parent_iv_score, social_iv_score]
Index: []

[0 rows x 26 columns]

There are no rows that do not contain the salary of the spouse if the spouse has a job. 

In order to fill in the missing values, we will need to have a standard feature for the annual fees (instead of having the fees feature expressed sometimes as fees per semester and other times as fees per year). We will then remove the fees and n_semesters from our dataset.

In [204]:
data['fees_per_annum'] = (data['fees'] / data['n_semesters']) * 2
data = data.drop(['fees', 'n_semesters'], axis= 1)

C:\Users\moham\miniconda3\envs\tensorflow\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Now, we will fill the missing values with reasonable assumptions.

If the fees/income is not specified, we fill it with fees_per_annum/annual_salary from the dataset. 

In [205]:
data['fees/income'] = data['fees/income'].fillna(data['fees_per_annum']/data['annual_salary'])

Same goes for spouse fees/income.

In [206]:
data['spouse_fees/income'] = data['spouse_fees/income'].fillna(data['fees_per_annum']/data['spouse_income'])

Filling the zeroes for the jobless spouses with zeroes.

In [207]:
data['spouse_income'] = data['spouse_income'].fillna(0)

In [208]:
data.isnull().any()

app_id                        False
job_name                       True
job_exp_yrs                   False
address                       False
house_type                    False
annual_salary                 False
salary_verification           False
prev_grade_percent            False
year                          False
fees/income                   False
down_pymnt                    False
spouse_has_job                False
total_balance                 False
payment_method                False
num_children                  False
credit_limit                  False
credit_limit_ex_mort          False
spouse_income                 False
spouse_fees/income             True
spouse_income_verification     True
decision                      False
mental_iv_score               False
parent_iv_score               False
social_iv_score               False
fees_per_annum                False
dtype: bool

The only feature with missing values now is spouse_fees/income because some spouses do not have a specified income. When aggregating later, we will deal with this feature.

After filling in the missing values, we will now see if some features should be eliminated from the dataset.

## **Feature Elimination**

Eliminated features and the rationale behind the elimination:

Applicant ID should have no significance on our model so we will remove it.

We believe salary and income verifications should have no significance on our model, so we will remove them assuming that all incomes or salaries are to be verified before enrolment of the applicant.

Year of application should also be removed since our model should not be affected by the time of the previous applications and also, the model will be used for future applicants where the year of application will have different values in the future.

Spouse_has_job, which indicates whether the spouse has a job or not, should be removed. Job details will be left if they exist and will be used in our model but we believe whether the spouse has a job or not should not affect the model.

Address will be removed as well because it has 50 unique variables. If address feature will stay, the values will need to be encoded and will generate many new features that will decrease the performance of our model.

Down payment also will be removed since it is paid after the decision of enrollment. So, it should not be used to predict the decision.

Also, job name should not affect the model, so we will keep other details about the job but we will remove the job name.

Finally, we believe payment method should not affect our model so we will remove it as well.

In [209]:
data = data.drop(['app_id', 'address', 'salary_verification', 'year', 'job_name', 'spouse_income_verification', 'payment_method', 'spouse_has_job', 'down_pymnt'], axis = 1)

Below is the data after the above features were eliminated.

In [210]:
data

job_exp_yrs house_type  annual_salary  prev_grade_percent  fees/income  \
0              1.0     rental        46000.0               92.65        17.48   
1              1.0     rental       152000.0               93.51         2.39   
2              5.0     rental       115000.0               88.47        12.95   
3              2.0     rental        25000.0               89.58         9.75   
4              6.0   mortgage        80000.0               78.51        15.17   
...            ...        ...            ...                 ...          ...   
33995          5.0     rental       105000.0               91.82        11.71   
33996          3.0   mortgage       118299.0               85.52        13.28   
33997          6.0   mortgage        56000.0               85.01        18.88   
33998          2.0   mortgage       107100.0               85.01         3.61   
33999          1.0   mortgage       125000.0               92.93         6.56   

       total_balance  num_children  credit_limit  credit_limit_ex_mort  \
0            35198.0             3       68039.0               35198.0   
1            11704.0             5       24000.0               11704.0   
2            41004.0             5       67122.0               41004.0   
3             4065.0             2       12000.0                4065.0   
4           169079.0             2      211158.0               32595.0   
...              ...           ...           ...                   ...   
33995        28742.0             2       60500.0               28742.0   
33996       330413.0             2      470072.0               69265.0   
33997       223575.0             4      279870.0               38038.0   
33998       178179.0             4      287972.0               17905.0   
33999       374885.0             5      432200.0               40925.0   

       spouse_income  spouse_fees/income      decision  mental_iv_score  \
0           110750.0               27.97  not enrolled         0.087638   
1                0.0                 NaN      enrolled        25.328459   
2                0.0                 NaN      enrolled         7.141356   
3                0.0                 NaN  not enrolled        34.474615   
4                0.0                 NaN      enrolled        29.070282   
...              ...                 ...           ...              ...   
33995            0.0                 NaN      enrolled        16.907394   
33996            0.0                 NaN      enrolled        65.635700   
33997            0.0                 NaN      enrolled        27.466363   
33998            0.0                 NaN      enrolled        24.455919   
33999            0.0                 NaN      enrolled        42.800443   

       parent_iv_score  social_iv_score  fees_per_annum  
0             8.423544        32.629941         80000.0  
1            12.940269        21.888056         40000.0  
2             7.539012        20.730058         20000.0  
3            22.732014         1.076913         40000.0  
4             5.226745        13.420388         54900.0  
...                ...              ...             ...  
33995        64.581258         1.609207         40000.0  
33996        24.085703         0.278597         24000.0  
33997        46.752626        15.781011         24000.0  
33998         3.753729        47.290821         70000.0  
33999        17.023625        25.215754         80000.0  

[33993 rows x 16 columns]

## **Further Exploration and Feature Engineering**

In [211]:
data.loc[(data['fees/income'] > 100)][['fees/income', 'decision', 'annual_salary', 'spouse_fees/income', 'spouse_income', 'fees_per_annum']]

fees/income      decision  annual_salary  spouse_fees/income  \
349         345.42      enrolled         6292.0               24.07   
872         220.91  not enrolled         5000.0               12.39   
1744        170.12      enrolled         8000.0               15.48   
5636        115.01      enrolled        10000.0               17.01   
5859        105.60  not enrolled        12000.0               37.16   
7408        131.35  not enrolled         5000.0               20.85   
8147        107.96      enrolled        36000.0               18.61   
8930        124.36  not enrolled        17000.0               27.44   
9487        146.33      enrolled        16500.0               27.75   
9627           inf      enrolled            0.0                4.45   
14860       185.81      enrolled         5500.0                9.25   
16050       167.06      enrolled        20000.0               12.48   
21087       109.76  not enrolled        20400.0               26.82   
21243       112.33      enrolled        27500.0               28.62   
21286     40000.00      enrolled            1.0                9.17   
21761          inf  not enrolled            0.0               19.62   
22142       740.96      enrolled         1000.0               12.28   
23483       214.43      enrolled         8400.0               25.16   
23738       222.95  not enrolled         7904.0               29.20   
24489       307.75      enrolled         3412.0               18.43   
24490       190.87      enrolled         2500.0               10.51   
25555       157.14      enrolled         5000.0               15.49   
25608       215.88      enrolled        10000.0               18.07   
25692       149.10      enrolled        10000.0               22.12   
27993       117.67  not enrolled        18000.0               23.31   
29046       260.24      enrolled         2000.0                5.34   
29114       248.50      enrolled        10000.0               22.96   

       spouse_income  fees_per_annum  
349         141292.0         64000.0  
872          89000.0         40000.0  
1744         98000.0         21000.0  
5636        115000.0         24000.0  
5859        107000.0         20400.0  
7408        285000.0         50000.0  
8147        306000.0         40000.0  
8930         77000.0         45350.0  
9487        136500.0         37000.0  
9627        112500.0         56000.0  
14860       166000.0         28000.0  
16050       642000.0         21000.0  
21087        83500.0         20000.0  
21243       267500.0         68950.0  
21286       100001.0         40000.0  
21761        85000.0         51700.0  
22142       121000.0         20000.0  
23483        80400.0         43850.0  
23738       108669.0         21000.0  
24489        70882.0         65200.0  
24490       332500.0         60000.0  
25555       265000.0         51250.0  
25608       141250.0         56000.0  
25692        84500.0         24000.0  
27993       138000.0         70000.0  
29046       552000.0         35000.0  
29114       110000.0         35000.0

As you can see, when the applicant's fees/income is very high (>100), their spouse always has a job and has a good salary & low fees/income ratio. Therefore, we can combine both the applicant & their spouse income to simplify the total household income. 

We will aggregate both parents' incomes and ratios for to have simplified aggregated features. We will add both incomes/salaries and calculate the new (fees/household_income) ratio.

We tried verifying the fees/ratio for both parents but the results in the dataset were not the same results that we got from our calculations. We believe this can be due to some non reported income sources or due to parents having other sources of income other than their salaries, like bank interests, private businesses, etc.

In [212]:
data['household_income'] = data['annual_salary'] + data['spouse_income']
data = data.drop(['annual_salary', 'spouse_income'], axis= 1)
data['fees/household_income'] = 1/(1/data['fees/income'] + 1/data['spouse_fees/income'])
data['fees/household_income'] = data['fees/household_income'].fillna(data['fees/income'])
data = data.drop(['fees/income', 'spouse_fees/income'], axis= 1)
data

job_exp_yrs house_type  prev_grade_percent  total_balance  \
0              1.0     rental               92.65        35198.0   
1              1.0     rental               93.51        11704.0   
2              5.0     rental               88.47        41004.0   
3              2.0     rental               89.58         4065.0   
4              6.0   mortgage               78.51       169079.0   
...            ...        ...                 ...            ...   
33995          5.0     rental               91.82        28742.0   
33996          3.0   mortgage               85.52       330413.0   
33997          6.0   mortgage               85.01       223575.0   
33998          2.0   mortgage               85.01       178179.0   
33999          1.0   mortgage               92.93       374885.0   

       num_children  credit_limit  credit_limit_ex_mort      decision  \
0                 3       68039.0               35198.0  not enrolled   
1                 5       24000.0               11704.0      enrolled   
2                 5       67122.0               41004.0      enrolled   
3                 2       12000.0                4065.0  not enrolled   
4                 2      211158.0               32595.0      enrolled   
...             ...           ...                   ...           ...   
33995             2       60500.0               28742.0      enrolled   
33996             2      470072.0               69265.0      enrolled   
33997             4      279870.0               38038.0      enrolled   
33998             4      287972.0               17905.0      enrolled   
33999             5      432200.0               40925.0      enrolled   

       mental_iv_score  parent_iv_score  social_iv_score  fees_per_annum  \
0             0.087638         8.423544        32.629941         80000.0   
1            25.328459        12.940269        21.888056         40000.0   
2             7.141356         7.539012        20.730058         20000.0   
3            34.474615        22.732014         1.076913         40000.0   
4            29.070282         5.226745        13.420388         54900.0   
...                ...              ...              ...             ...   
33995        16.907394        64.581258         1.609207         40000.0   
33996        65.635700        24.085703         0.278597         24000.0   
33997        27.466363        46.752626        15.781011         24000.0   
33998        24.455919         3.753729        47.290821         70000.0   
33999        42.800443        17.023625        25.215754         80000.0   

       household_income  fees/household_income  
0              156750.0              10.757219  
1              152000.0               2.390000  
2              115000.0              12.950000  
3               25000.0               9.750000  
4               80000.0              15.170000  
...                 ...                    ...  
33995          105000.0              11.710000  
33996          118299.0              13.280000  
33997           56000.0              18.880000  
33998          107100.0               3.610000  
33999          125000.0               6.560000  

[33993 rows x 14 columns]

Now, to be able to perform the correlation analysis, we will use one hot encoding to transform the categorical variables into numeric variables.

In [213]:
data.nunique()

job_exp_yrs                  9
house_type                   3
prev_grade_percent         350
total_balance            31406
num_children                 6
credit_limit             28631
credit_limit_ex_mort     27509
decision                     2
mental_iv_score          33993
parent_iv_score          33993
social_iv_score          33993
fees_per_annum            1137
household_income          3004
fees/household_income     4888
dtype: int64

House type and decision are the two categorical variables that we will need to transform. House type has 3 unique values while decision has 2 unique values. Therefore, house type will be encoded into two variables, rental and mortgage, while decision will be encoded into one variable, enrolled.

In [214]:
data['rental'] =  data['house_type'] == 'rental'
data['rental'] = data['rental'].astype(int)
data['mortgage'] =  data['house_type'] == 'mortgage'
data['mortgage'] = data['mortgage'].astype(int)

data['enrolled'] = data['decision'] == 'enrolled'
data['enrolled'] = data['enrolled'].astype(int)

data = data.drop(['house_type','decision'], axis= 1)
data

job_exp_yrs  prev_grade_percent  total_balance  num_children  \
0              1.0               92.65        35198.0             3   
1              1.0               93.51        11704.0             5   
2              5.0               88.47        41004.0             5   
3              2.0               89.58         4065.0             2   
4              6.0               78.51       169079.0             2   
...            ...                 ...            ...           ...   
33995          5.0               91.82        28742.0             2   
33996          3.0               85.52       330413.0             2   
33997          6.0               85.01       223575.0             4   
33998          2.0               85.01       178179.0             4   
33999          1.0               92.93       374885.0             5   

       credit_limit  credit_limit_ex_mort  mental_iv_score  parent_iv_score  \
0           68039.0               35198.0         0.087638         8.423544   
1           24000.0               11704.0        25.328459        12.940269   
2           67122.0               41004.0         7.141356         7.539012   
3           12000.0                4065.0        34.474615        22.732014   
4          211158.0               32595.0        29.070282         5.226745   
...             ...                   ...              ...              ...   
33995       60500.0               28742.0        16.907394        64.581258   
33996      470072.0               69265.0        65.635700        24.085703   
33997      279870.0               38038.0        27.466363        46.752626   
33998      287972.0               17905.0        24.455919         3.753729   
33999      432200.0               40925.0        42.800443        17.023625   

       social_iv_score  fees_per_annum  household_income  \
0            32.629941         80000.0          156750.0   
1            21.888056         40000.0          152000.0   
2            20.730058         20000.0          115000.0   
3             1.076913         40000.0           25000.0   
4            13.420388         54900.0           80000.0   
...                ...             ...               ...   
33995         1.609207         40000.0          105000.0   
33996         0.278597         24000.0          118299.0   
33997        15.781011         24000.0           56000.0   
33998        47.290821         70000.0          107100.0   
33999        25.215754         80000.0          125000.0   

       fees/household_income  rental  mortgage  enrolled  
0                  10.757219       1         0         0  
1                   2.390000       1         0         1  
2                  12.950000       1         0         1  
3                   9.750000       1         0         0  
4                  15.170000       0         1         1  
...                      ...     ...       ...       ...  
33995              11.710000       1         0         1  
33996              13.280000       0         1         1  
33997              18.880000       0         1         1  
33998               3.610000       0         1         1  
33999               6.560000       0         1         1  

[33993 rows x 15 columns]

## **Correlation Analysis**

In [215]:
main_corr_data = data.corr(method='pearson')
fig = px.imshow(main_corr_data,title="Merchants Pearson Correrlation Heatmap",width=1000,height=1000,labels={'color':"correlation"})
fig.show()

After performing correlation analysis, there are 2 very strong correlations that were noticed.

The first one is the correlation between total balance and credit limit. This is justifiable as banks tend to give higher credit limits to customers with higher balances. Because of this we will remove the credit limit feature from our dataset and leave only total balance.

The second correlation is the negative correlation between rental and mortgage. This is also justifiable since a house of type rental cannot be of type mortgage as well, and vice versa. However, we cannot remove one of the columns since we still have houses of type owned. Although the number of owned houses is less than the rented or mortgaged houses, but it is still a significant number and should not be classified as an outlier. Therefore, we keep both columns, rental and mortgage.

Moreover, there are moderate correlations noticed. A positive correlation between total balance and mortgage and a negative correlation between total balance and rental were noticed. This is also justifiable since mortgages are usually given to customers with high balances. However, we believe the correlation is not high enough to eliminate one of the columns.

Finally, we thought at the beginning that the 3 interview scores will have high correlations with the decision but this was not the case. The feature with the highest correlation with the decision is the applicant's previous grade. The higher the previous grade, the more probable the applicant is to get enrolled. This is the factor contributing the most to the enrollment decision.

In [216]:
# Remove the credit limit column as stated above.
data = data.drop(['credit_limit'], axis= 1)
data


job_exp_yrs  prev_grade_percent  total_balance  num_children  \
0              1.0               92.65        35198.0             3   
1              1.0               93.51        11704.0             5   
2              5.0               88.47        41004.0             5   
3              2.0               89.58         4065.0             2   
4              6.0               78.51       169079.0             2   
...            ...                 ...            ...           ...   
33995          5.0               91.82        28742.0             2   
33996          3.0               85.52       330413.0             2   
33997          6.0               85.01       223575.0             4   
33998          2.0               85.01       178179.0             4   
33999          1.0               92.93       374885.0             5   

       credit_limit_ex_mort  mental_iv_score  parent_iv_score  \
0                   35198.0         0.087638         8.423544   
1                   11704.0        25.328459        12.940269   
2                   41004.0         7.141356         7.539012   
3                    4065.0        34.474615        22.732014   
4                   32595.0        29.070282         5.226745   
...                     ...              ...              ...   
33995               28742.0        16.907394        64.581258   
33996               69265.0        65.635700        24.085703   
33997               38038.0        27.466363        46.752626   
33998               17905.0        24.455919         3.753729   
33999               40925.0        42.800443        17.023625   

       social_iv_score  fees_per_annum  household_income  \
0            32.629941         80000.0          156750.0   
1            21.888056         40000.0          152000.0   
2            20.730058         20000.0          115000.0   
3             1.076913         40000.0           25000.0   
4            13.420388         54900.0           80000.0   
...                ...             ...               ...   
33995         1.609207         40000.0          105000.0   
33996         0.278597         24000.0          118299.0   
33997        15.781011         24000.0           56000.0   
33998        47.290821         70000.0          107100.0   
33999        25.215754         80000.0          125000.0   

       fees/household_income  rental  mortgage  enrolled  
0                  10.757219       1         0         0  
1                   2.390000       1         0         1  
2                  12.950000       1         0         1  
3                   9.750000       1         0         0  
4                  15.170000       0         1         1  
...                      ...     ...       ...       ...  
33995              11.710000       1         0         1  
33996              13.280000       0         1         1  
33997              18.880000       0         1         1  
33998               3.610000       0         1         1  
33999               6.560000       0         1         1  

[33993 rows x 14 columns]

Above is the final data that will be fed into our predicitive models.

# **Predictive Models**

Now we need to convert the dataframe to a numpy array in order to feed the data to our predictive models. First we need to normalize our data.

In [219]:
normalized_data=(data-data.min())/(data.max()-data.min())

data_set = normalized_data.to_numpy()
print(data_set.shape)
data_set[:2]

(33993, 14)


array([[0.00000000e+00, 9.20560748e-01, 7.37509452e-03, 6.00000000e-01,
        3.92031944e-02, 9.89988862e-04, 9.54543143e-02, 3.67144768e-01,
        1.00000000e+00, 1.84140488e-02, 2.17757468e-01, 1.00000000e+00,
        0.00000000e+00, 0.00000000e+00],
       [0.00000000e+00, 9.54049844e-01, 2.45235827e-03, 1.00000000e+00,
        1.30358028e-02, 2.88305790e-01, 1.46637327e-01, 2.46277026e-01,
        3.33333333e-01, 1.77744383e-02, 4.83805668e-02, 1.00000000e+00,
        0.00000000e+00, 1.00000000e+00]])

In [220]:
X = data_set[:, 0:13]
y = data_set[:, 13]
(X_train, X_test, y_train, y_test) = train_test_split(X, y,
                                test_size=0.20, stratify=y, random_state=10)

print("Train size: ", X_train.shape)
print("Train Label size: ", y_train.shape)
print("Validation size: ", X_test.shape)
print("Validation Label size: ", y_test.shape)

Train size:  (27194, 13)
Train Label size:  (27194,)
Validation size:  (6799, 13)
Validation Label size:  (6799,)


In [ ]:
LR = 1e-4
epochs = 20
BS = 128
model = Sequential()
model.add(Dense(128, input_dim=13, activation='relu'))
model.add(Dense(64, activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(1, activation='sigmoid'))
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [86]:
H = model.fit(X_train, y_train, batch_size=BS,
             validation_data=(X_val, y_val),
             epochs=epochs)

Train on 27194 samples, validate on 6799 samples
Epoch 1/20
27194/27194 [==============================] - 2s 72us/step - loss: 0.5988 - accuracy: 0.6730 - val_loss: 0.5644 - val_accuracy: 0.7039
Epoch 2/20
27194/27194 [==============================] - 2s 83us/step - loss: 0.5620 - accuracy: 0.7088 - val_loss: 0.5627 - val_accuracy: 0.7088
Epoch 3/20
27194/27194 [==============================] - 1s 43us/step - loss: 0.5572 - accuracy: 0.7118 - val_loss: 0.5587 - val_accuracy: 0.7113
Epoch 4/20
27194/27194 [==============================] - 1s 52us/step - loss: 0.5559 - accuracy: 0.7128 - val_loss: 0.5595 - val_accuracy: 0.7107
Epoch 5/20
27194/27194 [==============================] - 2s 87us/step - loss: 0.5542 - accuracy: 0.7141 - val_loss: 0.5616 - val_accuracy: 0.7053
Epoch 6/20
27194/27194 [==============================] - 2s 63us/step - loss: 0.5544 - accuracy: 0.7116 - val_loss: 0.5597 - val_accuracy: 0.7116
Epoch 7/20
27194/27194 [==============================] - 2s 79us/ste

Now lets test the model on the testing set

In [221]:
file_name = "C:/Users/moham/OneDrive/Desktop/test_International-School-Dataset.csv"
#file_name = "C:/Users/HP/Downloads/train_International-School-Dataset.csv"
test_data = pd.read_csv(file_name)

Before feeding the testing set to any model, we will preprocess it such that it has the right dimentions & feature variables.

In [222]:
test_data = test_data[test_data['house_type'].isnull() == False]

test_data['fees_per_annum'] = (test_data['fees'] / test_data['n_semesters']) * 2
test_data = test_data.drop(['fees', 'n_semesters'], axis= 1)

test_data['fees/income'] = test_data['fees/income'].fillna(test_data['fees_per_annum']/test_data['annual_salary'])

test_data['spouse_fees/income'] = test_data['spouse_fees/income'].fillna(test_data['fees_per_annum']/test_data['spouse_income'])

test_data['spouse_income'] = test_data['spouse_income'].fillna(0)

test_data = test_data.drop(['app_id', 'address', 'salary_verification', 'year', 'job_name', 'spouse_income_verification', 'payment_method', 'spouse_has_job', 'down_pymnt'], axis = 1)

test_data['household_income'] = test_data['annual_salary'] + test_data['spouse_income']
test_data = test_data.drop(['annual_salary', 'spouse_income'], axis= 1)
test_data['fees/household_income'] = 1/(1/test_data['fees/income'] + 1/test_data['spouse_fees/income'])
test_data['fees/household_income'] = test_data['fees/household_income'].fillna(test_data['fees/income'])
test_data = test_data.drop(['fees/income', 'spouse_fees/income'], axis= 1)

test_data['rental'] =  test_data['house_type'] == 'rental'
test_data['rental'] = test_data['rental'].astype(int)
test_data['mortgage'] =  test_data['house_type'] == 'mortgage'
test_data['mortgage'] = test_data['mortgage'].astype(int)

test_data['enrolled'] = test_data['decision'] == 'enrolled'
test_data['enrolled'] = test_data['enrolled'].astype(int)

test_data = test_data.drop(['credit_limit'], axis= 1)

test_data = test_data.drop(['house_type','decision'], axis= 1)

test_data

job_exp_yrs  prev_grade_percent  total_balance  num_children  \
0             6.0               80.48       180643.0             3   
1             8.0               93.97       717111.0             3   
2             2.0               84.59        30109.0             3   
3             9.0               85.51       361935.0             4   
4             5.0               74.18        38835.0             5   
...           ...                 ...            ...           ...   
1995          5.0               86.41        10821.0             0   
1996          7.0               78.52       434063.0             3   
1997          6.0               92.51       624116.0             5   
1998          3.0               85.51       171689.0             4   
1999          3.0               76.01        27783.0             1   

      credit_limit_ex_mort  mental_iv_score  parent_iv_score  social_iv_score  \
0                  28558.0        23.649193         5.165801        17.779308   
1                  97434.0        19.830339        20.038635         6.336663   
2                  30109.0         1.339863         5.903078         9.462986   
3                  27013.0        21.294029         9.725985        44.028674   
4                  38835.0        17.226434        30.801950        19.968482   
...                    ...              ...              ...              ...   
1995               10821.0        15.141241         3.351711        29.615634   
1996               34762.0         6.001317         4.968623        15.215748   
1997               37491.0         2.645464        66.245158         4.956198   
1998               16419.0        18.686182         0.828035        23.097541   
1999               27783.0        16.973331        26.599551         4.877628   

      fees_per_annum  household_income  fees/household_income  rental  \
0            20000.0           80000.0                  12.49       0   
1            48000.0          156000.0                   3.33       0   
2            40000.0          107000.0                   7.26       1   
3            24000.0          100000.0                   7.35       0   
4            20000.0          300000.0                  10.38       1   
...              ...               ...                    ...     ...   
1995         45000.0           45000.0                   8.91       1   
1996         34475.0           76000.0                  11.38       0   
1997         42000.0          149500.0                   8.50       0   
1998         60000.0           86840.0                   8.97       1   
1999         22400.0           56000.0                  13.72       1   

      mortgage  enrolled  
0            1         0  
1            1         1  
2            0         0  
3            1         1  
4            0         0  
...        ...       ...  
1995         0         0  
1996         1         0  
1997         1         1  
1998         0         0  
1999         0         0  

[1999 rows x 14 columns]

In [223]:
normalized_test_data=(test_data-test_data.min())/(test_data.max()-test_data.min())

testing_set = normalized_test_data.to_numpy()

In [224]:
X_test = testing_set[:, 0:13]
y_test = testing_set[:, 13]

In [87]:
_, accuracy = model.evaluate(X_test, y_test)
print("Testing accuracy", accuracy*100)

1999/1999 [==============================] - 0s 165us/step
Testing accuracy 70.03501653671265


As you can see, this model achieved 70% accuracy on the testing set. Next we will be using a logistic regression model in order to compare their results.

In [81]:
model = LogisticRegression(solver='lbfgs', random_state=0).fit(X, y)
accuracy = model.score(X_test, y_test)
print("Testing accuracy", accuracy*100)

Testing accuracy 68.83441720860431


The logistic regression model achieved a 68.83% accuracy on the testing set. Slightlty less than the fully connected neural network model, but still very similar accuracy results. We suggest using the logistic regression model since it is lighter and needs less computing time with very similar results to the neural network model.